client/prometheus.go

```diff
@@ -0,0 +1,233 @@
+	increments         map[string]*incrementer.Prometheus
+}
+
+// NewPrometheus builds and returns new Prometheus instance
+func NewPrometheus(namespace string) (*Prometheus, error) {
+	client := &Prometheus{namespace: namespace, increments: make(map[string]*incrementer.Prometheus)}
+	return client, nil
+}
+
+// BuildTimer builds timer to track metric timings
+func (c *Prometheus) BuildTimer() timer.Timer {
+	return &timer.Memory{}
+}
+
+// Close closes underlying client connection if any
+func (c *Prometheus) Close() error {
+	return nil
+}
+
+// TrackRequest tracks HTTP Request stats
+func (c *Prometheus) TrackRequest(r *http.Request, t timer.Timer, success bool) Client {
+	// TODO
+	//b := bucket.NewHTTPRequest(c.httpRequestSection, r, success, c.httpMetricCallback, c.unicode)
+	//i := incrementer.NewPrometheus()
+	//
+	//t.Finish(b.Metric())
+	//i.IncrementAll(b)
+
+	return c
+}
+
+// TrackOperation tracks custom operation
+func (c *Prometheus) TrackOperation(section string, operation bucket.MetricOperation, t timer.Timer, success bool) Client {
+	// TODO
+	//b := bucket.NewPlain(section, operation, success, c.unicode)
+	//i := incrementer.NewPrometheus(map[string]string{"success": strconv.FormatBool(success)})
+	//
+	//if nil != t {
+	//	t.Finish(strings.Replace(b.MetricWithSuffix(), ".", "_", -1))
+	//}
+	//i.IncrementAll(b)
+	return c
+}
+
+// TrackOperationN tracks custom operation with n diff
+func (c *Prometheus) TrackOperationN(section string, operation bucket.MetricOperation, t timer.Timer, n int, success bool) Client {
+	// TODO
+	//b := bucket.NewPlainUnderline(section, operation, success, c.unicode)
+	//i := incrementer.NewPrometheus()
+	//
+	//if nil != t {
+	//	t.Finish(b.MetricWithSuffixUnderline())
+	//}
+	//i.IncrementAllN(b, n)
+
+	return c
+}
+
+// TrackMetric tracks custom metric, w/out ok/fail additional sections
+func (c *Prometheus) TrackMetric(section string, operation bucket.MetricOperation) Client {
+	b := bucket.NewPrometheus(section, operation, true, c.unicode)
+
+	metric := b.Metric()
+	metricTotal := b.MetricTotal()
+
+	if _, ok := c.increments[metric]; !ok {
+		i := incrementer.NewPrometheus()
+		i.Increment(metric)
+		c.increments[metric] = i
+	} else {
+		c.increments[metric].Increment(metric)
+	}
+
+	if _, ok := c.increments[metricTotal]; !ok {
+		i := incrementer.NewPrometheus()
+		i.Increment(metricTotal)
+		c.increments[metricTotal] = i
+	} else {
+		c.increments[metricTotal].Increment(metricTotal)
+	}
+
+	return c
+}
+
+// TrackMetric tracks custom metric with added labels, w/out ok/fail additional sections
+func (c *Prometheus) TrackMetricWithLabels(section string, operation bucket.MetricOperation, labels map[string]string) Client {
+	b := bucket.NewPrometheus(section, operation, true, c.unicode)
+
+	metric := b.Metric()
+	metricTotal := b.MetricTotal()
+
+	key_hash := ""
+	for key := range labels {
+		key_hash += "_" + key
+	}
+
+	if _, ok := c.increments[metric+key_hash]; !ok {
```

Comment:
you should use mutex for reading/writing to the map, otherwise app will panic on concurrent access